### ISSUE TITLE
Expose signing functionality?

### ISSUE THREAD

**Issue by jennybc** [paws-r/paws#428]
I have implemented an auth method called workload identity federation in the gargle R  package, which handles various matters related to calling Google APIs. Specifically, I have implemented this flow for AWS, i.e. for applications running on AWS that need to make authenticated calls to Google Cloud services.

As part of this token exchange ping pong, I have to craft a notional request to the AWS `GetCallerIdentity()` method. VERY IMPORTANTLY, I do not (can not) actual *make* this request. I need to marshal all the info, including the `x-amz-date` and `x-amz-security-token` headers, and generate a signature. Then I insert the signature into the headers, re-package the headers, serialize the whole request in a very specific way, and send it as part of a request to Google endpoint.

Here's the code:

https://github.com/r-lib/gargle/blob/0a2e7fd088c2ed52e088fe0d386f3d7adeac7d77/R/credentials_external_account.R#L264-L358

Currently I'm using `aws.signature::signature_v4_auth()` for generating the V4 signature.

But I've realized that, across various R packages maintained by RStudio folks, paws is a more common choice for AWS functionality. Therefore, for "meta" reasons, it makes sense for me to use paws as well.

However, I can't easily see that paws (paws.common?) exposes the signature functionality that I need. I can see it's here, for internal use.

Is there a way for me to do what I need with paws?

**Comment by davidkretch** [comment]
Hello! Sorry for the delayed response. Currently you're right -- that functionality is not exported. But I agree it should be.

Unfortunately I probably won't have time to do this until maybe later this month sadly. If you have time and want to address it before then, I'd be happy to accept it.

I think maybe a wrapper around [`sign_with_body`](https://github.com/paws-r/paws/blob/704379fece30a80fd186f4576a2c4eb0feb88c40/paws.common/R/signer_v4.R#L151) would work. The `signer` and `request` parameters expect a [signer](https://github.com/paws-r/paws/blob/704379fece30a80fd186f4576a2c4eb0feb88c40/paws.common/R/signer_v4.R#L62) object with credentials, and an [HTTP request](https://github.com/paws-r/paws/blob/704379fece30a80fd186f4576a2c4eb0feb88c40/paws.common/R/net.R#L6) with the URL.

Since you also need the instance region, there is also [get_instance_metadata](https://github.com/paws-r/paws/blob/704379fece30a80fd186f4576a2c4eb0feb88c40/paws.common/R/config.R#L161), which might work for you? Currently this is also not exported.

**Comment by jennybc** [comment]
OK glad to hear I read the situation correctly and that you think it's reasonable for paws.common to export such functionality.

Thanks for the pointers about how this might look. Not sure if I'll take a crack at a PR, but I might.

Yeah, I do also need to know the instance region.

**Comment by jennybc** [comment]
To provide additional color for whoever implements this in the end, here are some observations about what's awkward about the current approach using `aws.signature::signature_v4_auth()`. I am restating observations made in comments in the code I link to above.

The notional request must include the `x-amz-date` and `x-amz-security-token` headers. They both need to be among the signed headers. I would like to _not_ precompute those 2 headers (although I understand why it needs to be possible to pass them in), but *post hoc* I also need access to the values that were ultimately constructed and signed.

Basically, I would like the future requesting-signing function to return something that includes the entire set of signed headers, presumably the union of those I passed in and those computed internally, in the exact form used in the canonical request.
